packages/language-server/test/completion-items-utils.ts

```diff
@@ -0,0 +1,234 @@
+  return { document, position };
+}
+
+/** Use ⭲ to mark range start and ⭰ to mark range end */
+export function getRanges(xmlSnippet: string): Range[] {
+  const RANGE_START_CHAR = "⭲";
+  const RANGE_END_CHAR = "⭰";
+
+  const { document } = getXmlSnippetDocument(xmlSnippet);
+  const ranges: Range[] = [];
+  let xmlTextWithRanges = xmlSnippet.replace("⇶", "");
+  while (xmlTextWithRanges.indexOf(RANGE_START_CHAR) !== -1) {
+    const rangeStartIndex = xmlTextWithRanges.indexOf(RANGE_START_CHAR);
+    xmlTextWithRanges = xmlTextWithRanges.replace(RANGE_START_CHAR, "");
+    const rangeEndIndex = xmlTextWithRanges.indexOf(RANGE_END_CHAR);
+    if (rangeEndIndex === -1) {
+      break;
+    }
+    xmlTextWithRanges = xmlTextWithRanges.replace(RANGE_END_CHAR, "");
+    ranges.push({
+      start: document.positionAt(rangeStartIndex),
+      end: document.positionAt(rangeEndIndex)
+    });
+  }
+  return ranges;
+}
+
+export function getTextInRange(
+  xmlSnippet: string,
+  range: Range | undefined
+): string {
+  const { document } = getXmlSnippetDocument(xmlSnippet);
+  return document.getText(range);
+}
+
+function rangeToString(range: Range): string {
+  return `${range.start.line}:${range.start.character}-${range.end.line}:${range.end.character}`;
+}
+
+/** Check the suggestions will be displayed to the user according to the range and filter text */
+function assertSuggestionsAreValid(
+  suggestions: CompletionItem[],
+  xmlSnippet: string
+): void {
+  const { document, position } = getXmlSnippetDocument(xmlSnippet);
+  forEach(suggestions, suggestion => {
+    expectExists(suggestion.textEdit, "suggestion contains a textEdit");
+    assertRangeContains(suggestion.textEdit.range, position, suggestion.label);
+    assertRangesDoNotOverlap(
+      suggestion.label,
+      suggestion.textEdit,
+      suggestion.additionalTextEdits || []
+    );
+    // The filter text is checked until the position in the document
+    // (for example, we can replace "Ab⇶cd" with "Abzzz" even though "c" and "d" aren't in "Abzzz")
+    const checkedRange = {
+      start: suggestion.textEdit?.range.start,
+      end: position
+    };
+    assertFilterMatches(
+      suggestion.filterText ?? suggestion.label,
+      document.getText(checkedRange),
+      suggestion.label
+    );
+  });
+}
+
+function assertRangeContains(
+  range: Range,
+  position: Position,
+  description: string
+): void {
+  // The range must be in the same line as the position
+  expect(range.start.line, `${description}: range start line`).to.equal(
+    position.line
+  );
+  expect(range.end.line, `${description}: range end line`).to.equal(
+    position.line
+  );
+  expect(
+    range.start.character,
+    `${description}: range start character`
+  ).to.be.at.most(position.character);
+  expect(
+    range.end.character,
+    `${description}: range end character`
+  ).to.be.at.least(position.character);
+}
+
+function assertRangesDoNotOverlap(
```

Comment:
would this logic be much simpler if we **first** converted all these positions to offsets instead of line/colums?

Reply:
Done